judge/views/problem.py

```diff
@@ -521,128 +520,150 @@ def get(self, request, *args, **kwargs):
-        raise Http404()
-
-    if problem.is_editable_by(request.user):
-        judge_choices = tuple(Judge.objects.filter(online=True, problems=problem).values_list('name', 'name'))
-    else:
-        judge_choices = ()
-
-    if request.method == 'POST':
-        form = ProblemSubmitForm(request.POST, judge_choices=judge_choices,
-                                 instance=Submission(user=profile, problem=problem))
-        if form.is_valid():
-            if (not request.user.has_perm('judge.spam_submission') and
-                    Submission.objects.filter(user=profile, was_rejudged=False)
-                              .exclude(status__in=['D', 'IE', 'CE', 'AB']).count() >= settings.DMOJ_SUBMISSION_LIMIT):
-                return HttpResponse('<h1>You submitted too many submissions.</h1>', status=429)
-            if not problem.allowed_languages.filter(id=form.cleaned_data['language'].id).exists():
-                raise PermissionDenied()
-            if not request.user.is_superuser and problem.banned_users.filter(id=profile.id).exists():
-                return generic_message(request, _('Banned from submitting'),
-                                       _('You have been declared persona non grata for this problem. '
-                                         'You are permanently barred from submitting this problem.'))
-
-            with transaction.atomic():
-                if profile.current_contest is not None:
-                    contest_id = profile.current_contest.contest_id
-                    try:
-                        contest_problem = problem.contests.get(contest_id=contest_id)
-                    except ContestProblem.DoesNotExist:
-                        model = form.save()
-                    else:
-                        max_subs = contest_problem.max_submissions
-                        if max_subs and get_contest_submission_count(problem, profile,
-                                                                     profile.current_contest.virtual) >= max_subs:
-                            return generic_message(request, _('Too many submissions'),
-                                                   _('You have exceeded the submission limit for this problem.'))
-                        model = form.save()
-                        model.contest_object_id = contest_id
-
-                        contest = ContestSubmission(submission=model, problem=contest_problem,
-                                                    participation=profile.current_contest)
-                        contest.save()
-                else:
-                    model = form.save()
+    @cached_property
+    def contest_problem(self):
+        if self.request.profile.current_contest is None:
+            return None
+        return get_contest_problem(self.object, self.request.profile)
+
+    @cached_property
+    def remaining_submission_count(self):
+        max_subs = getattr(self.contest_problem, 'max_submissions', None)
```

Comment:
Apart from this stuff, this PR looks good, thanks.

Reply:
This has been changed in the latest commit.